Backports release 1.11

KristofferC opened the pull request:
Backported PRs:
- [x] #55773 <!-- Add compat entry for `Base.donotdelete` -->
- [x] #41244 <!-- Fix shell `cd` error when working dir has been deleted -->  
- [x] #55795 <!-- fix #52986, regression in `@doc` of macro without REPL loaded -->
- [x] #55829 <!-- [Dates] Make test more robust against non-UTC timezones --> 
- [x] #55641 <!-- fall back to slower stat filesize if optimized filesize fails -->
- [x] #55744 <!-- fix #45494, error in ssa conversion with complex type decl -->
- [x] #55783 <!-- use `inferencebarrier` instead of `invokelatest` for 1-arg `@assert` -->
- [x] #55739 <!-- Add `invokelatest` barrier to `string(...)` in `@assert` -->

Need manual backport:
- [ ] #55798 <!-- Broadcast binary ops involving strided triangular -->       

Contains multiple commits, manual intervention needed:
- [ ] #55509 <!-- Fix cong implementation to be properly random and not just cycling. -->
- [ ] #55569 <!-- Add a docs section about loading/precomp/ttfx time tuning -->
- [ ] #55824 <!-- Replace regex package module checks with actual code checks -->

Non-merged PRs with backport label:
- [ ] #55845 <!-- privatize annotated string API, take two -->
- [ ] #55828 <!-- Fix some corner cases of `isapprox` with unsigned integers -->
- [ ] #55813 <!-- Check for conflicting `@ccallable` name before JIT registration -->
- [ ] #55743 <!-- doc: heap snapshot viewing -->
- [ ] #55741 <!-- Change annotations to use a NamedTuple -->
- [ ] #55534 <!-- Set stdlib sources as read-only during installation -->     
- [ ] #55499 <!-- propagate the terminal's `displaysize` to the `IOContext` used by the REPL -->
- [ ] #55458 <!-- Allow for generically extracting unannotated string -->     
- [ ] #55457 <!-- Make AnnotateChar equality consider annotations -->
- [ ] #55453 <!-- Privatise the annotations API, for StyledStrings -->        
- [ ] #55355 <!-- relocation: account for trailing path separator in depot paths -->
- [ ] #55220 <!-- `isfile_casesensitive` fixes on Windows -->
- [ ] #55169 <!-- `propertynames` for SVD respects private argument -->       
- [ ] #54457 <!-- Make `String(::Memory)` copy -->
- [ ] #53957 <!-- tweak how filtering is done for what packages should be precompiled -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->
- [ ] #50813 <!-- More doctests for Sockets and capitalization fix -->        
- [ ] #50157 <!-- improve docs for `@inbounds` and `Base.@propagate_inbounds` -->

topolarity committed:
Add `invokelatest` barrier to `string(...)` in `@assert` (#55739)

This change protects `@assert` from invalidations to `Base.string(...)`
by adding an `invokelatest` barrier.

A common source of invalidations right now is `print(io,
join(args...))`. The problem is:
1. Inference concludes that `join(::Any...)` returns
`Union{String,AnnotatedString}`
2. The `print` call is union-split to `String` and `AnnotatedString`
3. This code is now invalidated when StyledStrings defines `print(io,
::AnnotatedString)`

The invalidation chain for `@assert` is similar: ` @assert 1 == 1` calls
into `string(::Expr)` which calls into `print(io, join(args::Any...))`.
Unfortunately that leads to the invalidation of almost all `@assert`s
without an explicit error message

Similar to
https://github.com/JuliaLang/julia/pull/55583#issuecomment-2308969806

(cherry picked from commit 945517ba4e15f7470b8790a696ba5404ef047f2f)

LilithHafner committed:
Add compat entry for `Base.donotdelete` (#55773)

(cherry picked from commit 346f38bceabf3dab1d3912fe822a663735c91d4a)

aviatesk committed:
use `inferencebarrier` instead of `invokelatest` for 1-arg `@assert` (#55783)

This version would be better as per this comment:
<https://github.com/JuliaLang/julia/pull/55739#pullrequestreview-2304360447>
I confirmed this still allows us to avoid invalidations reported at
JuliaLang/julia#55583.

(cherry picked from commit f8086062906d80ef56c0dcd595a13438ef028293)

norci committed:
Fix shell `cd` error when working dir has been deleted (#41244)

root cause:
if current dir has been deleted, then pwd() will throw an IOError:
pwd(): no such file or directory (ENOENT)

---------

Co-authored-by: Ian Butterworth <i.r.butterworth@gmail.com>
(cherry picked from commit 48ddd2dc2859f41fc1a82876cc5654fb31d7b2ba)

JeffBezanson committed:
fix #52986, regression in `@doc` of macro without REPL loaded (#55795)

fix #52986

(cherry picked from commit 4045e7baa496b2f14dfe562dbe3700d0c91006ff)

giordano committed:
[Dates] Make test more robust against non-UTC timezones (#55829)

`%M` is the format specifier for the minutes, not the month (which
should be `%m`), and it was used twice.

Also, on macOS `Libc.strptime` internally calls `mktime` which depends
on the local timezone. We now temporarily set `TZ=UTC` to avoid
depending on the local timezone.

Fix #55827.

(cherry picked from commit d6fa66ff6a983c08fce478346241879c7db31dce)

IanButterworth committed:
fall back to slower stat filesize if optimized filesize fails (#55641)

(cherry picked from commit fc9f1470458ad6bdeb1f56c1150d651814a0a164)

KristofferC committed:
privatize annotated string API, take two (#55845)

https://github.com/JuliaLang/julia/pull/55453 is stuck on StyledStrings
and Base documentation being entangled and there isn't a good way to
have the documentation of Base types / methods live in an stdlib. This
is a stop gap solution to finally be able to move forwards with 1.11.

(cherry picked from commit db6d277f6226daa5739940a4642277cfe0a884ea)

DilumAluthgeBot committed:
🤖 [master] Bump the Downloads stdlib from 1061ecc to 89d3c7d (#55854)

Stdlib: Downloads
URL: https://github.com/JuliaLang/Downloads.jl.git
Stdlib branch: master
Julia branch: master
Old commit: 1061ecc
New commit: 89d3c7d
Julia version: 1.12.0-DEV
Downloads version: 1.6.0(It's okay that it doesn't match)
Bump invoked by: @KristofferC
Powered by:
[BumpStdlibs.jl](https://github.com/JuliaLang/BumpStdlibs.jl)

Diff:
https://github.com/JuliaLang/Downloads.jl/compare/1061ecc377a053fce0df94e1a19e5260f7c030f5...89d3c7dded535a77551e763a437a6d31e4d9bf84

```
$ git log --oneline 1061ecc..89d3c7d
89d3c7d fix cancelling upload requests (#259)
df33406 gracefully cancel a request (#256)
```

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>
(cherry picked from commit c3af4fc24564c3ecda59a26648abe919f090929a)

aviatesk commented:
The cherry-picked commit for the parser fix seems to be broken? Currently this branch doesn't bootstrap:
```
➜ make -j8 && jlb -e 'using Pkg; Pkg.precompile()' && jlb --startup-file=no test/runtests.jl compiler staged reflection REPL/replcompletions opaque_closure
    LINK usr/lib/libjulia-codegen.1.11.dylib
    LINK usr/lib/libjulia-codegen.dylib
    JULIA usr/lib/julia/corecompiler.ji
error during bootstrap:
LoadError(at "compiler/compiler.jl" line 3: LoadError(at "compiler/optimize.jl" line 219: LoadError(at "compiler/ssair/passes.jl" line 1565: ErrorException("syntax: malformed expression"))))
jl_f_throw at ~/julia/julia5/src/builtins.c:579
jl_apply at ~/julia/julia5/src/julia.h:2157 [inlined]
do_call at ~/julia/julia5/src/interpreter.c:126
eval_stmt_value at ~/julia/julia5/src/interpreter.c:174
eval_body at ~/julia/julia5/src/interpreter.c:663
jl_interpret_toplevel_thunk at ~/julia/julia5/src/interpreter.c:821
top-level scope at compiler/ssair/passes.jl:1565
jl_toplevel_eval_flex at ~/julia/julia5/src/toplevel.c:943
...
```

KristofferC commented:
I dropped it

nsajko commented:
Added a PR to update NEWS.md: #55858

nsajko committed:
NEWS entry for deletion of `length(::Stateful)` (#55858)

Will update History.md on master if this gets merged.

xref #47790

xref #51747

xref #54953

---------

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
